### PR TITLE
Allow spacetimedb to run in an in-memory mode

### PR DESCRIPTION
**Pull request by kulakowski**
# Description of Changes

# API

 - [ ] This is a breaking change to the module API
 - [ ] This is a breaking change to the ClientAPI
 - [ ] This is a breaking change to the SDK API

*If the API is breaking, please state below what will break*

### CODE REVIEW

**Review by Centril — COMMENTED** [review]
Minor readability stuff

File: crates/bench/src/spacetime.rs [hunk 1]
```diff
@@ -50,7 +50,9 @@ impl BuildDb for DbResult {
     where
         Self: Sized,
     {
-        let db = build_db()?;
+        // For benchmarking, we are concerned with the persistent version of the database.
+        let in_memory = false;
```
[the remark this answers is not in the document]
Comment (mamcx) [reply crates/bench/src/spacetime.rs:54]: Or better like `enum Stored {Memory, Disk}`?

File: crates/core/src/db/ostorage/memory_object_db.rs [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+    objects: HashMap<Hash, Bytes>,
+}
+
+impl MemoryObjectDB {
+    pub fn new() -> Self {
+        eprintln!("GGKK in memory");
+        let objects = HashMap::new();
+        Self { objects }
+    }
+}
```
[the remark this answers is not in the document]
Comment (kulakowski) [reply crates/core/src/db/ostorage/memory_object_db.rs:19]: Ah yeah. GGKK is my initials doubled up. Unique and greppable, and professional (ie not just cursing).

Done.

File: crates/standalone/src/subcommands/start.rs [hunk 3]
```diff
@@ -14,7 +14,38 @@ impl From<std::string::String> for OsStr {
     }
 }
 
-pub fn cli(is_standalone: bool) -> clap::Command {
+pub enum ProgramKind {
```
[the remark this answers is not in the document]
Comment (kulakowski) [reply crates/standalone/src/subcommands/start.rs:17]: Done.

File: crates/standalone/src/subcommands/start.rs [hunk 4]
```diff
@@ -14,7 +14,38 @@ impl From<std::string::String> for OsStr {
...
+    Standalone,
+    CLI,
+}
+
+impl ProgramKind {
+    fn listen_addr(&self) -> &'static str {
```
[the remark this answers is not in the document]
Comment (kulakowski) [reply crates/standalone/src/subcommands/start.rs:23]: Done (and also the one below).

File: crates/bench/src/spacetime.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (kulakowski) [reply crates/bench/src/spacetime.rs:54]: Done.

**Review by Centril — APPROVED** [review]

**Comment by kulakowski** [comment]
Don't land this yet.

One of the other bits of mutable storage (presumably either the control db or the object store for the wasm blobs) has been relying on creating the message log creating some intermediate directory as a side effect. A small fix, but gimme a minute.

**Comment by kulakowski** [comment]
> Don't land this yet.
> 
> One of the other bits of mutable storage (presumably either the control db or the object store for the wasm blobs) has been relying on creating the message log creating some intermediate directory as a side effect. A small fix, but gimme a minute.

Resolved.

@cloutiertyler tested as well

File: crates/standalone/src/subcommands/start.rs [hunk 5]
```diff
@@ -88,15 +123,21 @@ pub fn cli(is_standalone: bool) -> clap::Command {
...
-        // we want the executable name to be `spacetime` when we're executing this from the CLI. We have to
-        // pass these strings with static lifetimes so we can't do any dynamic string manipulation here.
-        .after_help(if is_standalone {
-            "Run `spacetimedb help start` for more detailed information."
-        } else {
-            "Run `spacetime help start` for more information."
-        })
+        .arg(jwt_priv_key_path_arg);
+
+    // Standalone databases have the option to run entirely in memory.
```
Comment (jdetter) [inline crates/standalone/src/subcommands/start.rs:128]: @kulakowski  Why is this standalone only? Can we not also add this functionality to the `spacetime` command as well?

Comment (kulakowski) [reply crates/standalone/src/subcommands/start.rs:128]: Fixed by https://github.com/clockworklabs/SpacetimeDB/pull/148

**Pull request merged**